api/provider/aws/test_aws/environment_read_test.go

```diff
@@ -0,0 +1,104 @@
+package test_aws
+
+import (
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/golang/mock/gomock"
+	provider "github.com/quintilesims/layer0/api/provider/aws"
+	"github.com/quintilesims/layer0/api/tag"
+	awsc "github.com/quintilesims/layer0/common/aws"
+	"github.com/quintilesims/layer0/common/config/mock_config"
+	"github.com/quintilesims/layer0/common/models"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestEnvironmentRead(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	// todo: setup helper for config
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+
+	tags := models.Tags{
+		{
+			EntityID:   "env_id",
+			EntityType: "environment",
+			Key:        "name",
+			Value:      "env_name",
+		},
+		{
+			EntityID:   "env_id",
+			EntityType: "environment",
+			Key:        "os",
+			Value:      "linux",
+		},
+	}
+
+	for _, tag := range tags {
+		if err := tagStore.Insert(tag); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// environments' security group names are <fq environment id>-env
+	describeSecurityGroupHelper(mockAWS, "l0-test-env_id-env", "sg_id")
+
+	// environments' asg names are the same as the fq environment id
+	describeASGInput := &autoscaling.DescribeAutoScalingGroupsInput{}
+	describeASGInput.SetAutoScalingGroupNames([]*string{aws.String("l0-test-env_id")})
+
+	// environments use the asg.LaunchConfigurationName as the source of truth for
+	// their launch configuration name
+	asg := &autoscaling.Group{}
+	asg.SetAutoScalingGroupName("l0-test-env_id")
+	asg.SetLaunchConfigurationName("lc_name")
+	asg.SetInstances(make([]*autoscaling.Instance, 2))
+
+	describeASGOutput := &autoscaling.DescribeAutoScalingGroupsOutput{}
+	describeASGOutput.SetAutoScalingGroups([]*autoscaling.Group{asg})
+
+	mockAWS.AutoScaling.EXPECT().
+		DescribeAutoScalingGroups(describeASGInput).
+		Return(describeASGOutput, nil)
+
+	describeLCInput := &autoscaling.DescribeLaunchConfigurationsInput{}
+	describeLCInput.SetLaunchConfigurationNames([]*string{aws.String("lc_name")})
+
+	lc := &autoscaling.LaunchConfiguration{}
+	lc.SetLaunchConfigurationName("lc_name")
+	lc.SetInstanceType("m3.small")
+	lc.SetImageId("ami_id")
+
+	describeLCOutput := &autoscaling.DescribeLaunchConfigurationsOutput{}
+	describeLCOutput.SetLaunchConfigurations([]*autoscaling.LaunchConfiguration{lc})
+
+	mockAWS.AutoScaling.EXPECT().
+		DescribeLaunchConfigurations(describeLCInput).
+		Return(describeLCOutput, nil)
+
+	target := provider.NewEnvironmentProvider(mockAWS.Client(), tagStore, mockConfig)
+	result, err := target.Read("env_id")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// todo: Links
+	expected := &models.Environment{
```

Comment:
Would it be impractical to define `expected` near the top of the test function and use fields from that model while setting up the rest of the test?

Reply:
It's definitely not impractical, but there is a school of thought that it's better to use hardcoded values over variables/references during unit tests (within reason, of course). Reason being: you actually _don't_ want it to be too easy to change your unit tests once they have been created - if someone introduces a breaking change, you want whoever did it to see all of the (potentially) cascading effects. I also tend to favor that approach because it aligns with Go's tendency towards explicitness. 

I don't feel very strong either way on the matter, but wanted to expose the logic behind why I wrote it in this form.

Reply:
OK, that's a good argument; I was looking at it strictly from an ease-of-writing perspective. If the pattern sits well with everyone else, fine by me.